### PR TITLE
polling for status changes in admin and waiting pages

### PR DESCRIPTION
**Pull request by thehamzarocks**

### CODE REVIEW

File: simplq/src/components/pages/Admin/index.jsx [hunk 1]
```diff
@@ -25,14 +25,20 @@ export default () => {
 
   const [items, setItems] = useState();
 
-  const update = () => {
+  const update = (updateClicked) => {
     if (queueId) {
       QueueService.get(queueId)
         .then((data) => {
           setItems(data.tokens);
+          if (!updateClicked) {
```
Comment (thehamzarocks) [inline simplq/src/components/pages/Admin/index.jsx:33]: Using updateClicked to determine if this is a call from polling or from a button click might be messy. Shall I just create a separate function pollForUpdates instead of using if's everywhere?

File: simplq/src/components/pages/QueueStatus.jsx [hunk 2]
```diff
@@ -64,10 +77,6 @@ function QueueStatus() {
       });
   };
 
-  if (aheadCount == null && !updateInProgress) {
```
Comment (thehamzarocks) [inline simplq/src/components/pages/QueueStatus.jsx:67]: I presume we don't need this anymore

File: simplq/src/components/pages/Admin/index.jsx [hunk 3]
```diff
@@ -62,7 +68,7 @@ export default () => {
         queueId={queueId}
         className={styles.shareButton}
         onRefresh={() => {
-          update();
+          update(true);
           setItems(false);
```
Comment (thehamzarocks) [inline simplq/src/components/pages/Admin/index.jsx:72]: This seems redundant. Why was this added? Can it be removed safely?

Comment (thehamzarocks) [reply simplq/src/components/pages/Admin/index.jsx:72]: I believe it can

Comment (daltonfury42) [reply simplq/src/components/pages/Admin/index.jsx:72]: Do you want to completely rewrite this page? not the HTML/Styling part, but the logic?

Comment (thehamzarocks) [reply simplq/src/components/pages/Admin/index.jsx:72]: Hmm ok. I'll take it up. And the queustatus file too

File: simplq/src/components/pages/Admin/index.jsx (hunk 1 above)
Comment (maaverik) [reply simplq/src/components/pages/Admin/index.jsx:33]: Yes, please. It'll be cleaner.

File: simplq/src/components/pages/Admin/index.jsx [hunk 4]
```diff
@@ -37,6 +37,22 @@ export default () => {
...
+        .then((data) => {
+          setItems(data.tokens);
+          setTimeout(pollAndUpdate, 10000);
+        })
+        .catch((err) => {
+          handleApiErrors(err);
+          setTimeout(pollAndUpdate, 10000);
+        });
+    }
+  };
```
Comment (daltonfury42) [inline simplq/src/components/pages/Admin/index.jsx:52]: This is a very interesting problem. I think we can have a single function that takes care of both. Then the function can be called at start using `useEffect` at start and also when the button is clicked.

Use `clearTimeout` to cancel the previous timeout when the function is called again. Can you give this a try? The thing to figure out is where to store the `timeoutID`

File: simplq/src/components/pages/QueueStatus.jsx (hunk 2 above)
Comment (daltonfury42) [reply simplq/src/components/pages/QueueStatus.jsx:67]: shitty code.. 😓

File: simplq/src/components/pages/Admin/index.jsx (hunk 4 above)
Comment (thehamzarocks) [reply simplq/src/components/pages/Admin/index.jsx:52]: A recommended way involves using a higher-order component called timeout which has timeout methods and passing in those methods as props to our child component.

An alternative way would be to use the component's state.

File: simplq/src/components/pages/Admin/index.jsx [hunk 5]
```diff
@@ -24,19 +24,26 @@ export default () => {
...
   const [items, setItems] = useState();
+  const [timeoutId, setTimeoutId] = useState();
 
   const update = () => {
+    clearTimeout(timeoutId);
     if (queueId) {
       QueueService.get(queueId)
         .then((data) => {
           setItems(data.tokens);
+          const timeout = setTimeout(update, 10000);
```
Comment (daltonfury42) [inline simplq/src/components/pages/Admin/index.jsx:35]: declare this as const on top: 10000

**Comment by maaverik** [comment]
If someone approves, all subsequent commits on the PR don't need approval to merge. This is not how it should be, right? Is there some setting for this?

File: simplq/src/components/pages/Admin/index.jsx [hunk 6]
```diff
@@ -26,17 +38,24 @@ export default () => {
   const [items, setItems] = useState();
 
   const update = () => {
+    clearTimeout(timeoutUtil.getTimeoutId());
     if (queueId) {
       QueueService.get(queueId)
         .then((data) => {
           setItems(data.tokens);
+          const timeoutId = setTimeout(update, 10000);
```
Comment (thehamzarocks) [inline simplq/src/components/pages/Admin/index.jsx:46]: could make it a single-liner

File: simplq/src/components/pages/QueueStatus.jsx [hunk 7]
```diff
@@ -37,21 +49,25 @@ function QueueStatus() {
   const classes = useStyles();
 
   const update = () => {
+    clearTimeout(timeoutUtil.getTimeoutId());
     if (tokenId) {
-      setUpdateInProgress(true);
```
Comment (thehamzarocks) [inline simplq/src/components/pages/QueueStatus.jsx:41]: We don't want to show the half-second spinner every 10s

File: simplq/src/components/pages/Admin/index.jsx [hunk 8]
```diff
@@ -26,17 +28,27 @@ export default () => {
...
         });
     }
   };
 
+  useEffect(() => {
+    update();
+    return () => clearTimeout(timeoutId);
+  }, [queueId]);
+
+  useEffect(update, [queueId]);
```
Comment (daltonfury42) [inline simplq/src/components/pages/Admin/index.jsx:50]: We don't need two? First one should take care?

Comment (thehamzarocks) [reply simplq/src/components/pages/Admin/index.jsx:50]: Yup let me remove the second

**Comment by daltonfury42** [comment]
> If someone approves a PR once, all subsequent commits on the PR don't need approval to merge. This is not how it should be, right? Is there some setting for this?

image

Yes, the dismiss stale requests option should be unchecked. I think you've already done that

**Comment by daltonfury42** [comment]
@thehamzarocks nit: can you declare 10000 as a constant on top?

**Comment by maaverik** [comment]
I see this warning after the new useEffect was added:

.```
/src/components/pages/Admin/index.jsx
  Line 49:6:  React Hook useEffect has a missing dependency: 'update'. Either include it or remove the dependency array  react-hooks/exhaustive-deps

./src/components/pages/QueueStatus.jsx
  Line 61:6:  React Hook useEffect has a missing dependency: 'update'. Either include it or remove the dependency array  react-hooks/exhaustive-deps

Search for the keywords to learn more about each warning.
To ignore, add // eslint-disable-next-line to the line before.
```

It points to the useEffect, i.e.,
 ```
 useEffect(() => {
    update();
    return () => clearTimeout(timeoutId);
  }, [tokenId]);
```

Is it okay to remove the dependencies tokenId and queueId from the useEffects or should I suppress the warning?

**Comment by thehamzarocks** [comment]
I believe removing tokenId from useEffect will make the effect run on every render, which is not what we need.
It's better to suppress warnings or use useCallBack to memoize the function:

https://stackoverflow.com/questions/55840294/how-to-fix-missing-dependency-warning-when-using-useeffect-react-hook
